### PR TITLE
EX-1552 - Implement parser to get MGS filesystems

### PR DESCRIPTION
**Pull request by johnsonw**
Implement a new parser in lustre_collector that can take the output of lctl get_param -N mgs.MGS.live.* and parse it to show filesystem layouts

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

### CODE REVIEW

**Comment by johnsonw** [comment]
After compiling and deploying to the mgs node here is some sample output from a test cluster:

**TLDR**
```
{
    "Target": {
      "FsNames": {
        "kind": "MGT",
        "param": "fsnames",
        "target": "MGS",
        "value": [
          "fs"
        ]
      }
    }
  },
```

**Full Output**:
```
[
  {
    "Host": {
      "Memused": {
        "param": "memused",
        "value": 403538759
      }
    }
  },
  {
    "Host": {
      "MemusedMax": {
        "param": "memused_max",
        "value": 404026330
      }
    }
  },
  {
    "Host": {
      "LNetMemUsed": {
        "param": "lnet_memused",
        "value": 106089321
      }
    }
  },
  {
    "Host": {
      "HealthCheck": {
        "param": "health_check",
        "value": "healthy"
      }
    }
  },
  {
    "Target": {
      "ConnectedClients": {
        "kind": "MDT",
        "param": "connected_clients",
        "target": "fs-MDT0001",
        "value": 0
      }
    }
  },
  {
    "Target": {
      "Stats": {
        "kind": "MGT",
        "param": "stats",
        "target": "MGS",
        "value": [
          {
            "name": "req_waittime",
            "units": "usec",
            "samples": 37877,
            "min": 11,
            "max": 2695,
            "sum": 6180400,
            "sumsquare": 1268908618
          },
          {
            "name": "req_qdepth",
            "units": "reqs",
            "samples": 37877,
            "min": 0,
            "max": 1,
            "sum": 76,
            "sumsquare": 76
          },
          {
            "name": "req_active",
            "units": "reqs",
            "samples": 37877,
            "min": 1,
            "max": 2,
            "sum": 44289,
            "sumsquare": 57113
          },
          {
            "name": "req_timeout",
            "units": "sec",
            "samples": 37877,
            "min": 1,
            "max": 10,
            "sum": 37886,
            "sumsquare": 37976
          },
          {
            "name": "reqbuf_avail",
            "units": "bufs",
            "samples": 103909,
            "min": 62,
            "max": 64,
            "sum": 6544469,
            "sumsquare": 412244641
          },
          {
            "name": "ldlm_plain_enqueue",
            "units": "reqs",
            "samples": 201,
            "min": 1,
            "max": 1,
            "sum": 201,
            "sumsquare": 201
          },
          {
            "name": "mgs_connect",
            "units": "usec",
            "samples": 9,
            "min": 52,
            "max": 5165,
            "sum": 19362,
            "sumsquare": 66639088
          },
          {
            "name": "mgs_disconnect",
            "units": "usec",
            "samples": 4,
            "min": 50,
            "max": 92,
            "sum": 265,
            "sumsquare": 18709
          },
          {
            "name": "mgs_target_reg",
            "units": "usec",
            "samples": 90,
            "min": 874,
            "max": 163383,
            "sum": 1262544,
            "sumsquare": 91852108168
          },
          {
            "name": "mgs_config_read",
            "units": "usec",
            "samples": 41,
            "min": 41,
            "max": 2203,
            "sum": 26823,
            "sumsquare": 32448779
          },
          {
            "name": "obd_ping",
            "units": "usec",
            "samples": 36936,
            "min": 3,
            "max": 4398,
            "sum": 1901345,
            "sumsquare": 160282185
          },
          {
            "name": "llog_origin_handle_open",
            "units": "usec",
            "samples": 153,
            "min": 29,
            "max": 16443,
            "sum": 25516,
            "sumsquare": 270992222
          },
          {
            "name": "llog_origin_handle_next_block",
            "units": "usec",
            "samples": 298,
            "min": 24,
            "max": 31952,
            "sum": 141030,
            "sumsquare": 2788155300
          },
          {
            "name": "llog_origin_handle_read_header",
            "units": "usec",
            "samples": 145,
            "min": 25,
            "max": 44125,
            "sum": 192095,
            "sumsquare": 4905765639
          }
        ]
      }
    }
  },
  {
    "Target": {
      "ThreadsMax": {
        "kind": "MGT",
        "param": "threads_max",
        "target": "MGS",
        "value": 32
      }
    }
  },
  {
    "Target": {
      "ThreadsMin": {
        "kind": "MGT",
        "param": "threads_min",
        "target": "MGS",
        "value": 3
      }
    }
  },
  {
    "Target": {
      "ThreadsStarted": {
        "kind": "MGT",
        "param": "threads_started",
        "target": "MGS",
        "value": 4
      }
    }
  },
  {
    "Target": {
      "NumExports": {
        "kind": "MGT",
        "param": "num_exports",
        "target": "MGS",
        "value": 5
      }
    }
  },
  {
    "Target": {
      "ContendedLocks": {
        "kind": "MDT",
        "param": "contended_locks",
        "target": "fs-MDT0001",
        "value": 32
      }
    }
  },
  {
    "Target": {
      "ContentionSeconds": {
        "kind": "MDT",
        "param": "contention_seconds",
        "target": "fs-MDT0001",
        "value": 2
      }
    }
  },
  {
    "Target": {
      "CtimeAgeLimit": {
        "kind": "MDT",
        "param": "ctime_age_limit",
        "target": "fs-MDT0001",
        "value": 10
      }
    }
  },
  {
    "Target": {
      "EarlyLockCancel": {
        "kind": "MDT",
        "param": "early_lock_cancel",
        "target": "fs-MDT0001",
        "value": 0
      }
    }
  },
  {
    "Target": {
      "LockCount": {
        "kind": "MDT",
        "param": "lock_count",
        "target": "fs-MDT0001",
        "value": 0
      }
    }
  },
  {
    "Target": {
      "LockTimeouts": {
        "kind": "MDT",
        "param": "lock_timeouts",
        "target": "fs-MDT0001",
        "value": 0
      }
    }
  },
  {
    "Target": {
      "LockUnusedCount": {
        "kind": "MDT",
        "param": "lock_unused_count",
        "target": "fs-MDT0001",
        "value": 0
      }
    }
  },
  {
    "Target": {
      "LruMaxAge": {
        "kind": "MDT",
        "param": "lru_max_age",
        "target": "fs-MDT0001",
        "value": 3900000
      }
    }
  },
  {
    "Target": {
      "LruSize": {
        "kind": "MDT",
        "param": "lru_size",
        "target": "fs-MDT0001",
        "value": 800
      }
    }
  },
  {
    "Target": {
      "MaxNolockBytes": {
        "kind": "MDT",
        "param": "max_nolock_bytes",
        "target": "fs-MDT0001",
        "value": 0
      }
    }
  },
  {
    "Target": {
      "MaxParallelAst": {
        "kind": "MDT",
        "param": "max_parallel_ast",
        "target": "fs-MDT0001",
        "value": 1024
      }
    }
  },
  {
    "Target": {
      "ResourceCount": {
        "kind": "MDT",
        "param": "resource_count",
        "target": "fs-MDT0001",
        "value": 0
      }
    }
  },
  {
    "Target": {
      "Stats": {
        "kind": "MDT",
        "param": "md_stats",
        "target": "fs-MDT0001",
        "value": [
          {
            "name": "statfs",
            "units": "reqs",
            "samples": 8164,
            "min": null,
            "max": null,
            "sum": null,
            "sumsquare": null
          }
        ]
      }
    }
  },
  {
    "Target": {
      "NumExports": {
        "kind": "MDT",
        "param": "num_exports",
        "target": "fs-MDT0001",
        "value": 11
      }
    }
  },
  {
    "Target": {
      "FilesFree": {
        "kind": "MDT",
        "param": "filesfree",
        "target": "fs-MDT0001",
        "value": 2083568
      }
    }
  },
  {
    "Target": {
      "FilesTotal": {
        "kind": "MDT",
        "param": "filestotal",
        "target": "fs-MDT0001",
        "value": 2083840
      }
    }
  },
  {
    "Target": {
      "BytesAvail": {
        "kind": "MDT",
        "param": "kbytesavail",
        "target": "fs-MDT0001",
        "value": 2678816768
      }
    }
  },
  {
    "Target": {
      "BytesFree": {
        "kind": "MDT",
        "param": "kbytesfree",
        "target": "fs-MDT0001",
        "value": 2943983616
      }
    }
  },
  {
    "Target": {
      "BytesTotal": {
        "kind": "MDT",
        "param": "kbytestotal",
        "target": "fs-MDT0001",
        "value": 2945957888
      }
    }
  },
  {
    "Target": {
      "FsNames": {
        "kind": "MGT",
        "param": "fsnames",
        "target": "MGS",
        "value": [
          "fs"
        ]
      }
    }
  },
  {
    "LNetStat": {
      "SendCount": {
        "nid": "0@lo",
        "param": "send_count",
        "value": 33838
      }
    }
  },
  {
    "LNetStat": {
      "RecvCount": {
        "nid": "0@lo",
        "param": "recv_count",
        "value": 33836
      }
    }
  },
  {
    "LNetStat": {
      "DropCount": {
        "nid": "0@lo",
        "param": "drop_count",
        "value": 2
      }
    }
  },
  {
    "LNetStat": {
      "SendCount": {
        "nid": "10.73.20.12@tcp",
        "param": "send_count",
        "value": 471877
      }
    }
  },
  {
    "LNetStat": {
      "RecvCount": {
        "nid": "10.73.20.12@tcp",
        "param": "recv_count",
        "value": 471819
      }
    }
  },
  {
    "LNetStat": {
      "DropCount": {
        "nid": "10.73.20.12@tcp",
        "param": "drop_count",
        "value": 40
      }
    }
  }
]
```

File: src/mgs/mgs_parser.rs [hunk 1]
```diff
@@ -126,7 +169,52 @@ where
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use combine::{many, parser::EasyParser};
+    use insta::assert_debug_snapshot;
+
+    #[test]
+    fn test_params() {
```
Comment (jgrund) [inline src/mgs/mgs_parser.rs:190]: There should ideally be three tests here:

1. Test output with a single mounted fs
1. Test output with multiple mounted fs
1. Test output with no mounted fs

In addition, the tests should be smaller to just test the new parser in isolation.

That's not to say this larger test needs to be removed as it's useful as well.

Comment (johnsonw) [reply src/mgs/mgs_parser.rs:190]: Sure. I didn't see a test for the mgs so I figured I would write a test that included all of the output. I'll add a couple of tests to cover the cases listed above.

File: src/main.rs [hunk 2]
```diff
@@ -54,7 +64,12 @@ fn main() {
 
     let handle = thread::spawn(move || -> Result<Vec<Record>, LustreCollectorError> {
         let lctl_output = get_lctl_output()?;
-        let lctl_record = parse_lctl_output(&lctl_output)?;
+        let lctl_name_only_output = get_lctl_name_only_output()?;
```
Comment (jgrund) [inline src/main.rs:67]: I'm not convinced having a separate parser just for the name only is the best way to handle this.

File: src/mgs/mgs_parser.rs [hunk 3]
```diff
@@ -34,13 +35,40 @@ pub fn params() -> Vec<String> {
     .collect::<Vec<_>>()
 }
 
+pub fn name_only_params() -> Vec<String> {
+    vec!["mgs.MGS.live.*".to_string()]
```
Comment (jgrund) [inline src/mgs/mgs_parser.rs:39]: What does this look like if there are multiple MGS?

Comment (johnsonw) [reply src/mgs/mgs_parser.rs]: I'm using a wildcard character instead of `MGS` now.

File: src/main.rs (hunk 2 above)
Comment (johnsonw) [reply src/main.rs]: This has been updated to instead use a new module and pull this information using the modules parser in its own thread.

File: src/mgs/mgs_parser.rs (hunk 1 above)
Comment (johnsonw) [reply src/mgs/mgs_parser.rs:141]: This has been updated in the new module.

File: src/mgs/mgs_fs_parser.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (mgs_fs_stat())
+        .map(|(param, stat)| match stat {
+            MgsFsStat::MgsFsNames(xs) => {
+                let mgs_map: HashMap<Target, Vec<FsName>> =
+                    xs.into_iter()
+                        .fold(HashMap::new(), |mut acc, (target, fs_name)| {
+                            if let Some(fs_names) = acc.get(&target) {
```
Comment (jgrund) [inline src/mgs/mgs_fs_parser.rs:78]: nit, this would probably be cleaner by using the entry api.

File: src/parser.rs [hunk 5]
```diff
@@ -29,6 +29,7 @@ where
         top_level_parser::parse().map(|x| vec![x]),
         client_count_parser::parse(),
         mgs_parser::parse().map(|x| vec![x]),
+        mgs_fs_parser::parse(),
```
Comment (jgrund) [inline src/parser.rs:32]: Does this belong here?

File: src/mgs/mgs_fs_parser.rs (hunk 4 above)
Comment (johnsonw) [reply src/mgs/mgs_fs_parser.rs:78]: cool thanks for the tip. I updated to use this instead.

File: src/parser.rs (hunk 5 above)
Comment (johnsonw) [reply src/parser.rs]: Nope, removed.

File: src/mgs/mgs_fs_parser.rs [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (
+        attempt(string("mgs")).skip(period()),
+        target().skip(period()),
+        string("live").skip(period()),
+        word().skip(newline()).map(FsName),
+    )
+        .map(|(_, target, _, fsname)| (target, fsname))
```
Comment (ip1981) [inline src/mgs/mgs_fs_parser.rs:39]:
```suggestion
        attempt(string("mgs")).skip(period()).with(target().skip(period())),
        string("live").skip(period()).with(word().skip(newline()).map(FsName)),
    )

```

Just a note :)

Comment (johnsonw) [reply src/mgs/mgs_fs_parser.rs:39]: Sure, either way, works :)

**Pull request merged**